src/Interpreters/Cache/LockedKey.cpp

```diff
@@ -0,0 +1,161 @@
+#include <Interpreters/Cache/LockedKey.h>
+#include <Interpreters/Cache/FileSegment.h>
+#include <Interpreters/Cache/LockedFileCachePriority.h>
+#include <Interpreters/Cache/FileCache.h>
+#include <Common/logger_useful.h>
+#include <filesystem>
+
+namespace fs = std::filesystem;
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+LockedKey::LockedKey(
+    const FileCacheKey & key_,
+    KeyMetadata & key_metadata_,
+    KeyGuard::Lock && lock_,
+    const std::string & key_path_,
+    KeysQueuePtr cleanup_keys_metadata_queue_)
+    : key(key_)
+    , key_path(key_path_)
+    , lock(std::move(lock_))
+    , key_metadata(key_metadata_)
+    , cleanup_keys_metadata_queue(cleanup_keys_metadata_queue_)
+    , log(&Poco::Logger::get("LockedKey"))
+{
+}
+
+LockedKey::~LockedKey()
+{
+    removeKeyIfEmpty();
+}
+
+bool LockedKey::isLastOwnerOfFileSegment(size_t offset) const
+{
+    const auto * file_segment_metadata = key_metadata.getByOffset(offset);
+    return file_segment_metadata->file_segment.use_count() == 2;
+}
+
+void LockedKey::removeFileSegment(
+    size_t offset,
+    const FileSegmentGuard::Lock & segment_lock,
+    const CacheGuard::Lock & cache_lock)
+{
+    LOG_DEBUG(
+        log, "Remove from cache. Key: {}, offset: {}",
+        key.toString(), offset);
+
+    auto * file_segment_metadata = key_metadata.getByOffset(offset);
+
+    if (file_segment_metadata->queue_iterator)
+        LockedCachePriorityIterator(cache_lock, file_segment_metadata->queue_iterator).remove();
+
+    const auto cache_file_path = file_segment_metadata->file_segment->getPathInLocalCache();
+    file_segment_metadata->file_segment->detach(segment_lock, *this);
+
+    key_metadata.erase(offset);
```

Comment:
Actually I think that try catch it redundant, it was a precaution, I'd better remove it.
But anyway, the answer is no, we should erase it from the metadata, because isReleasable was satisfied for this file segment.

Reply:
Removed.